demos/signature/rsa_pss_direct.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/bio.h>
+#include "rsa_pss.h"
+
+/*
+ * The digest to be signed. This should be the output of a hash function.
+ * Here we sign an all-zeroes digest for demonstration purposes.
+ */
+static const unsigned char test_digest[32] = {0};
```

Comment:
If you use an initializer list, C specifies that any unspecified elements are zero-initialized.

(Because of this, you'd expect `= {}` to be reasonable, but the C standard is weird about this; technically you have to specify at least one value, so `= {0}`. In practice most compilers will accept `= {}`, but e.g. MSVC in C mode won't. IIRC C2x might be set to finally allow `= {}`.)

(Plus, since this is of static storage duration, it would be zero-initialized even without the initializer list here; it's technically unnecessary, and included for exposition).

From C11:
>The initialization shall occur in initializer list order, each initializer provided for a particular subobject overriding any previously listed initializer for the same subobject; all subobjects that are not initialized explicitly shall be initialized implicitly the same as objects that have static storage duration.

Reply:
Yes, but we use C89. I'm not sure this is true in that version.

Reply:
From C89:

>If there are fewer initializers in a list than there are members of an aggregate, the remainder of the aggregate shall be initialized implicitly the same as objects that have static storage duration.

Reply:
Ok.